File: solutions/zig/perfect-numbers/1/perfect_numbers.zig
const std = @import("std");

pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

fn sumFactors(n: u64) u64 {
    var sum: u64 = 0;
    for (1..(n / 2 + 1)) |candidate| {
        if (n % candidate == 0) {
            sum += candidate;
        }
    }
    return sum;
}

pub fn classify(n: u64) Classification {
    std.debug.assert(n != 0);
    const sum = sumFactors(n);
    if (sum > n) {
        return Classification.abundant;
    } else if (sum < n) {
        return Classification.deficient;
    } else {
        return Classification.perfect;
    }
}

File: solutions/elixir/perfect-numbers/1/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(1), do: {:ok, :deficient}

  def classify(number) when number > 0 do
    aliquot_sum = 1..div(number, 2)
    |> Enum.filter(&(rem(number, &1) == 0))
    |> Enum.sum

    diff = number - aliquot_sum

    class = cond do
      diff > 0 -> :deficient
      diff < 0-> :abundant
      :else -> :perfect
    end
    {:ok, class}
  end

  def classify(_number)  do
    {:error, "Classification is only possible for natural numbers."}
  end


end
